bindings/rust/evmc-declare/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+                        "evm1" => 0x1u32,
+                        "ewasm" => 0x1u32 << 1u32,
+                        _ => panic!("Invalid capabilities specifier. Use 'evm1' or 'ewasm'."),                       
+                    },
+                    _ => panic!("Argument 1 is not a valid string literal."),
+                };
+                (name, capabilities)
+            } else {
+                panic!("Argument passed to evmc_declare_vm is not a name-value item.")
+            }
+        }
+        _ => panic!("Argument passed to evmc_declare_vm is of incorrect type. Expected a name-value item.")
+    };
+    
+    // Get the VM version from the crate version.
+    let vm_version_string = env!("CARGO_PKG_VERSION").to_string();
+
+    // Get all the tokens from the respective helpers.
+    let static_data_tokens =
+        build_static_data(&vm_name_stylized, &vm_name_allcaps, &vm_version_string);
+    let container_tokens = build_vm_container();
+    let capabilities_tokens = build_capabilities_fn(&vm_name_lowercase, vm_capabilities);
+    let create_tokens = build_create_fn(&vm_name_lowercase, &vm_name_allcaps, &vm_type_name);
+    let destroy_tokens = build_destroy_fn(&vm_name_lowercase, &vm_type_name);
+    let execute_tokens = build_execute_fn(&vm_name_lowercase, &vm_type_name);
+
+    let quoted = quote! {
+        #input
+        #static_data_tokens
+        #container_tokens
+        #capabilities_tokens
+        #create_tokens
+        #destroy_tokens
+        #execute_tokens
+    };
+
+    quoted.into()
+}
+
+fn build_execute_fn(name_lowercase: &String, type_name: &String) -> proc_macro2::TokenStream {
+    let fn_name_string = format!("{}_execute", name_lowercase);
+    let fn_name_ident = Ident::new(&fn_name_string, name_lowercase.span());
+    let type_name_ident = Ident::new(type_name, type_name.span());
+
+    quote! {
+        extern "C" fn #fn_name_ident(
+            instance: *mut ::evmc_sys::evmc_instance,
+            context: *mut ::evmc_sys::evmc_context,
+            rev: ::evmc_sys::evmc_revision,
+            msg: *const ::evmc_sys::evmc_message,
+            code: *const u8,
+            code_size: usize
+        ) -> ::evmc_sys::evmc_result
+        {
+            assert!(!msg.is_null());
+            assert!(!context.is_null());
+            assert!(!instance.is_null());
+            assert!(!code.is_null());
+
+            let execution_context = ::evmc_vm::ExecutionContext::new(
+                msg.as_ref().expect("EVMC message is null"),
+                context.as_mut().expect("EVMC context is null")
+            );
+
+            let code_ref: &[u8] = unsafe {
+                ::std::slice::from_raw_parts(code, code_size);
+            }
+
+            let container = __EvmcContainer::from_ffi_pointer::<#type_name_ident>(instance);
+
+            let result = container.execute(code_ref, &execution_context);
+
+            container.into_ffi_pointer();
```

Comment:
Not sure why shuffling this one struct around is needed? Also `from_ffi_pointer` seems to create a clone first?

Reply:
Was there an answer to this or the code is gone (haven't checked)?

Reply:
It is not cloned anymore.